content/develop/concepts/app-design/multithreading.md

```diff
@@ -0,0 +1,180 @@
+
+In the following example, five custom threads are created from the script thread. After the threads are finished running, their results are displayed in the app.
+
+```python
+import streamlit as st
+import time
+from threading import Thread
+
+
+class WorkerThread(Thread):
+    def __init__(self, delay):
+        super().__init__()
+        self.delay = delay
+        self.return_value = None
+
+    def run(self):
+        start_time = time.time()
+        time.sleep(self.delay)
+        end_time = time.time()
+        self.return_value = f"start: {start_time}, end: {end_time}"
+
+
+delays = [5, 4, 3, 2, 1]
+threads = [WorkerThread(delay) for delay in delays]
+for thread in threads:
+    thread.start()
+for thread in threads:
+    thread.join()
+for i, thread in enumerate(threads):
+    st.header(f"Thread {i}")
+    st.write(thread.return_value)
+
+st.button("Rerun")
+```
+
+<Cloud name="doc-multithreading-no-st-commands-batched" height="700px" />
+
+If you want to display results in your app as various custom threads finish running, use containers. In the following example, five custom threads are created similarly to the previous example. However, five containers are initialized before running the custom threads and a `while` loop is used to display results as they become available. Since the Streamlit `write` command is called outside of the custom threads, this does not raise an exception.
+
+```python
+import streamlit as st
+import time
+from threading import Thread
+
+
+class WorkerThread(Thread):
+    def __init__(self, delay):
+        super().__init__()
+        self.delay = delay
+        self.return_value = None
+
+    def run(self):
+        start_time = time.time()
+        time.sleep(self.delay)
+        end_time = time.time()
+        self.return_value = f"start: {start_time}, end: {end_time}"
+
+
+delays = [5, 4, 3, 2, 1]
+result_containers = []
+for i, delay in enumerate(delays):
+    st.header(f"Thread {i}")
+    result_containers.append(st.container())
+
+threads = [WorkerThread(delay) for delay in delays]
+for thread in threads:
+    thread.start()
+thread_lives = [True] * len(threads)
+
+while any(thread_lives):
+    for i, thread in enumerate(threads):
+        if thread_lives[i] and not thread.is_alive():
+            result_containers[i].write(thread.return_value)
+            thread_lives[i] = False
+    time.sleep(0.5)
+
+for thread in threads:
+    thread.join()
+
+st.button("Rerun")
+```
+
+<Cloud name="doc-multithreading-no-st-commands-iterative" height="700px" />
+
+### Option 2: Expose `ScriptRunContext` to the thread
+
+If you want to call Streamlit commands from within your custom threads, you must attach the correct `ScriptRunContext` to the thread.
+
+<Warning>
+
+- This is not officially supported and may change in a future version of Streamlit.
+- This may not work with all Streamlit commands.
```

Comment:
Do we know exhaustively which commands these might be, or should this just be a generic warning to cover the possibilities?